fplll/pruner/pruner.h

```diff
@@ -290,47 +391,158 @@ template <class FT> class Pruner
+    }
     load_basis_shapes(gso_rs);
   }
 
   /**
-     Run the optimization process, successively using the algorithm activated
+     @brief Main interface for optimizing pruning coefficients.
+
+     It will invoke either of the two functions:
+
+     1. `optimize_coefficients_cost_vary_prob()` or
+     2. `optimize_coefficients_cost_fixed_prob()`
+
+     depending on the input flags.
 
-     @brief run the optimization process
+     If the flag PRUNER_SINGLE is disabled (default), it calls function (1) so that goal is to
+     optimize the cost computed as`single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0)`
+
+     If the flag PRUNER_SINGLE is enabled in `flags`, it calls function (2) such that the goal is to
+     optimize the single_enum_cost(pr) while fixing the metric (number of solutions or success
+     probability) to the target.
   */
   void optimize_coefficients(/*io*/ vector<double> &pr);
 
-  /** @brief Compute the cost of a single enumeration */
+  /**
+     @brief Main interface for optimizing the pruning coefficients with respect to the overall
+     enumeraiton time.
+
+     Main interface to optimize the overall enumeration time where the target function is:
+     single_enum_cost(pr) * trials + preproc_cost * (trials - 1.0);
+
+     Hierarchy of calls: This function first invokes optimize_coefficients_evec() which optimizes
+     using only even-position vectors for speed. Then it calls several local tuning functions
+     optimize_coefficients_local_adjust_*() to adjust the parameters in small scales. Finally it
+     does an optimization using full vectors using optimize_coefficients_full(). This procedure is
+     repeated for several rounds until no further improvements can be achieved.
+  */
+  void optimize_coefficients_cost_vary_prob(/*io*/ vector<double> &pr);
 
+  /**
+     @brief Main interface to optimize the pruning coefficients with repeect to the single enumeration.
+
+     Main interface to optimize the single enumeration time with the constraint such that the succ.
+     prob (or expected solutions) is fixed (and given) from input to the Pruner constructor.
+
+     Hierarchy of calls: This function first invokes optimize_coefficients_evec() and then
+     optimize_coefficients_full() to optimize the overall enumeration cost. Then it tries to adjust
+     the pruning parameters to achieve the target succ. probability (or expected number of
+     solutions) by calling either optimize_coefficients_incr_prob() or
+     optimize_coefficients_decr_prob(). Finally, it does some local optimization by calling
+     optimize_coefficients_local_adjust_smooth() which aims to smooth the discountinuities in the
+     curve and then optimize_coefficients_local_adjust_prob() which aims to fine-adjust the succ.
+     probability to be close enough to the target.
+  */
+  void optimize_coefficients_cost_fixed_prob(/*io*/ vector<double> &pr);
```

Comment:
`optimize_coefficients_cost_fixed_prob` Sorry to be on the names again, but `prob` is only one metric, should this be `fixed_metric`, also should we drop the `_cost_`?

Reply:
Maritn, I consider this more a constraint optimization problem, e.g.
optimize_coefficients_TargetFunction_Constraint

Currently we have 2 of them:
optimize_coefficients_cost_fixed_prob
optimize_coefficients_cost_fixed_prob

Presumably we may also have  optimize_coefficients_prob_fixed_cost in the future. In such case, the current naming convention seems easier to extend?